### ISSUE TITLE
Object-level animations

### ISSUE THREAD

**Issue by fireclawthefox** [panda3d/panda3d#457]
Panda3D currently lacks object-level animations which could be handy for a variety of tasks. One for example to make it more easy to animate collision solids created in a 3D modelling application and exported to an egg file.

To add support for them, the following parts probably have to be extended
- The Animation system
- Egg model format
- Exporters for 3D modelling apps

**Comment by rdb** [comment]
I agree, this is important to me; it would be a great boon to be able to properly define eg. door-open animations in the modelling program.  Doing it using vertex animation is a hack that has major caveats, such as not causing collision solids to be animated, as you point out.

This could be useful for doors, levers or really any objects that move rigidly in an environment.  But also think about cutscenes, which define an animation for multiple objects in a scene at the same time.  It would be great to be able to create animated cutscenes in Blender and then play those back in Panda3D.

One possible implementation is to add a `bind_anim` method on NodePath that binds a loaded AnimBundle object directly to a tree of nodes, instead of to a tree of joints.  It would recursively bind different animation channels to different sub-nodes according to name and hierarchy, and return an AnimControl using which you can play and stop the animation.  Under the hood, it would assign a new type of RenderEffect (eg. AnimationEffect) to each animated node that makes sure to update the position of the node according to its bound channel.

That would not be hard to implement; a drawback of that approach is that blending multiple animations at the same time would not be automatically supported the way it is for skeletal animations.  That said, I'm not sure that I can think of a use case for animation blending on object animations.

Another complication is that binding an animation by node and hierarchy might be a bit fragile for complex node trees.  I am not sure that there is a great solution to this—other than simply protecting the node from being flattened—if we want it to remain possible to mix-and-match animations (possibly loaded from different files) and apply animations to multiple nodes at a time.

I'd be quite happy for others to weigh in on this.

**Comment by wezu** [comment]
Well you sort of can right now. One _just_ needs to add a joint/bone/armature and and hard rig a model to it (set all weights to exactly 1.0) and animate that joint as if it was the actual object. For collisions - there's `actor.exposeJoint()`.

**Comment by rdb** [comment]
Yeah, but it's not very efficient (Panda will move all the individual vertices instead of changing the object transform), requires some awkward extra setup and doesn't work if you want to move an object that already has a skeletal animation.

A better way you could do currently is to create an armature with an empty (or single-vertex) mesh attached and exposing its one joint, and then parenting the object to the exposed joint.  It's still a bit clunky, but more efficient because Panda isn't animating any vertices.  I suppose we could set up YABEE and the .egg loader to support this approach; that would not require any new APIs.  We could even have the .egg loader automatically create such a dummy joint for every character to avoid the need for changes to the .egg structure.

I'm not really sure how well that approach would work when you have many objects in the scene, all with their own animations, though.  There is no way for the `auto_bind` mechanism to be able to know which animation belongs to which node other than by matching the name exactly.  Plus, Actor only supports one Character for an egg file.  I see several ways to deal with that:
* Load each animated object separately, as its own Actor.  Requires a lot of custom code on the Panda side, though, and makes scene-wide animations (eg. cutscenes) difficult.
* Load the entire scene as one big actor, with different animations acting on different parts, something like:
```python
scene = Actor('scene.egg', {'door3-open': 'open-door3-anim.egg'})
scene.play('door3-open')
```
This would be easiest for cutscenes, which move a lot of objects at once, but it seems like it would be an absolute pain to do for every individual moving object in the scene.
* We could change Actor to be able to load egg files containing multiple Characters as a multipart actor automatically.  You could then define one animation that opens all doors but play it on specific parts if desired.  So to play animation 'door-open' but affecting 'door3' only, you could do:
```python
scene = Actor('scene.egg', {'door-open': 'open-all-doors-anim.egg'}, autoMultipart=True)
scene.play('door-open', partName='door3')
```
This would support both cutscenes and individual animations well, at least for the Blender workflow.

My own thought is that all of these are bit annoying because they all require creating an Actor around animated parts, but what do you guys feel?  For comparison, this is what a new API similar to what I proposed earlier could look like:
```python
scene = loader.loadModel('scene.egg')
door = scene.find('**/door3')
ctrl = door.bindAnim(loader.loadAnim('door-open-anim.egg'))
ctrl.play()
```

**Comment by Epihaius** [comment]
As a developer of an editor that might/could/should support animation, I would definitely benefit from an easy-to-use object-level animation system as well.

Until now I had been considering using LerpIntervals for this purpose, but exporting those would likely be problematic anyway.

As for the proposed solutions, if the requirement of creating an Actor can be avoided then I'm all for it.
It will already be challenging enough to procedurally create animation using the existing API, so the NodePath.bind_anim implementation would certainly get my vote at this point.
